src/mscorlib/src/System/Reflection/Emit/CustomAttributeBuilder.cs

```diff
@@ -150,12 +156,26 @@ private bool ValidateType(Type t)
             // Now verify that the types of the actual parameters are compatible with the types of the formal parameters.
             for (i = 0; i < paramTypes.Length; i++)
             {
-                if (constructorArgs[i] == null)
+                object constructorArg = constructorArgs[i];
+                if (constructorArg == null)
+                {
+                    if (paramTypes[i].IsPrimitive)
+                    {
+                        throw new ArgumentNullException("constructorArgs[" + i + "]");
+                    }
                     continue;
+                }
+                Type constructorArgType = constructorArg.GetType();
                 TypeCode paramTC = Type.GetTypeCode(paramTypes[i]);
-                if (paramTC != Type.GetTypeCode(constructorArgs[i].GetType()))
-                    if (paramTC != TypeCode.Object || !ValidateType(constructorArgs[i].GetType())) 
+                if (paramTC != Type.GetTypeCode(constructorArgType))
+                {
+                    if (paramTC != TypeCode.Object || !ValidateType(constructorArgType))
                         throw new ArgumentException(Environment.GetResourceString("Argument_BadParameterTypeForConstructor", i));
+                }
+                if (constructorArgType == typeof(IntPtr) || constructorArgType == typeof(UIntPtr))
+                {
+                    throw new ArgumentException(Environment.GetResourceString("Argument_BadParameterTypeForConstructor", i), "constructorArgs[" + i + "]");
```

Comment:
This is necessary in the case of passing an `IntPtr` or `UIntPtr` to an `object` parameter:
E.g.
```

public class Program
{
    static unsafe void Main(string[] args)
    {
        IntPtr i = (IntPtr)1;
        CustomAttributeBuilder c = new CustomAttributeBuilder(typeof(HAttribute).GetConstructors()[0], new object[] { i });
        AssemblyBuilder assembly = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("Name"), AssemblyBuilderAccess.Run);
        assembly.SetCustomAttribute(c);
        // CustomAttributeFormatException thrown!
        assembly.GetCustomAttributes();

        Console.ReadLine();
    }
}

public class HAttribute : Attribute
{
    public HAttribute(object o) { }
}
```